### ISSUE TITLE
ChatCompletionClient to support request caching

### ISSUE THREAD

**Issue by ekzhu** [microsoft/autogen#4752]
Support client-side caching for any `ChatCompletionClient` type.

Simplest way to do it is to create a `ChatCompletionCache` type that implements the `ChatCompletionClient` protocol but wraps an existing client. 

Example how this may work:

```python
from autogen_ext.stores.diskcache import DiskCacheStore
from autogen_ext.models.cache import ChatCompletionCache
from autogen_ext.models.openai import OpenAIChatCompletionClient

# Cached client.
cached_client = ChatCompletionCache(OpenAIChatCompletionClient(model="gpt-4o"), store=DiskCacheStore())
```

**Comment by srjoglekar246** [comment]
Heres a basic idea I have based on [what we had](https://github.com/microsoft/autogen/tree/0.2/autogen/cache) in `0.2`:

1. We add `AbstractStoreBase` as the primary interface in `autogen_core` for cache stores.
2. We implement `InMemoryStore` etc in `autogen_ext` along with a general `Store` factory.
3. This will allow the implementation of a Cached Client interface similar to what Eric mentioned above:

```
    from autogen_ext.store.in_memory_store import InMemoryStore
    model_client = OpenAIChatCompletionClient(
        model="gpt-4o",
        api_key="...",
    )
    model_client = ChatCompletionCache(model_client, InMemoryStore())
    
    print(f"Model info: {model_client.model_info}")
    print("\n")
    
    prompt_messages = [
        SystemMessage(content=SYSTEM_PROMPT),
        UserMessage(content=USER_PROMPT, source="user"),
    ]
    
    num_prompt_tokens = model_client.count_tokens(prompt_messages)
    print(f"Prompt tokens: {num_prompt_tokens}")
    
    result = await model_client.create(messages=prompt_messages)
    print(f"create output: {result.content}")
```

We can modify the `Result` instances from this class to have `cached=True` etc.

As for the actual caching, since we use pydantic Models for the messages we can encode the incoming prompt info as a json & hash it for the cache key.

WDYT @ekzhu / @jackgerrits ?

**Comment by ekzhu** [comment]
Sounds good to me. For the abstract interface we can keep it super simple so existing libraries like `diskcache` and `redis` can just duck-type it. e.g., an interface with just `set` and `get`.

**Comment by rickyloynd-microsoft** [comment]
On a related note, for cases where the user requires all responses to be pulled from the cache, such as for quick regression tests, it could be useful to have the cached client throw an error (rather than calling the model_client) for any prompt that is not found in the cache. This functionality could be enabled by passing None as the model_client parameter. I've implemented a client wrapper that provides this caching and checking (plus numeric result checking) for my own regression tests, but my client wrapper isn't a complete ChatCompletionClient replacement.

**Comment by srjoglekar246** [comment]
@rickyloynd-microsoft Can you share a pointer/branch to your code, if possible?

**Comment by rickyloynd-microsoft** [comment]
It will be in a PR soon.

**Comment by victordibia** [comment]
I see the api we have documented here using context [here](cached_client = ChatCompletionCache(OpenAIChatCompletionClient(model="gpt-4o"), store=DiskCacheStore()))

```python
async def main() -> None:
    with tempfile.TemporaryDirectory() as tmpdirname: 
        openai_model_client = OpenAIChatCompletionClient(model="gpt-4o")  
        cache_store = DiskCacheStore[CHAT_CACHE_VALUE_TYPE](Cache(tmpdirname))
        cache_client = ChatCompletionCache(openai_model_client, cache_store) 
        response = await cache_client.create([UserMessage(content="Hello, how are you?", source="user")])
        print(response)  # Should print response from OpenAI
        response = await cache_client.create([UserMessage(content="Hello, how are you?", source="user")])
        print(response)  # Should print cached response
```

Does the current impl support something like

```python
async def main() -> None: 
        cache_client = OpenAIChatCompletionClient(model="gpt-4o", cache=True) 
        response = await cache_client.create([UserMessage(content="Hello, how are you?", source="user")])
        print(response)  # Should print response from OpenAI
        response = await cache_client.create([UserMessage(content="Hello, how are you?", source="user")])
        print(response)  # Should print cached response
```

@ekzhu @gagb

**Comment by gagb** [comment]
Nice simplification @victordibia !

**Comment by srjoglekar246** [comment]
@victordibia I suppose that can be done. What would be a good default cache? in memory?

**Comment by jackgerrits** [comment]
See my other reply: https://github.com/microsoft/autogen/issues/5141#issuecomment-2610897935

I believe the code snippet you showed breaks the intentional generality of the new cache client. We should avoid it being possible to cache multiple ways.

How does this snippet look?

```py
cached_model_client = ChatCompletionCache(OpenAIChatCompletionClient(model="gpt-4o"))
```

We just need to define a default cache (I agree in memory is good) and it's doable. IMO it makes sense and is a better design as it is entirely separated from a specific client and works for *all* at once.

**Comment by victordibia** [comment]
Makes sense to have something that is general and seperated from the chatcompletion implementation itself. 
The API you have seems like a good compromise, assuming `cached_model_client ` will be a drop-in replacement for `model_client`. In this case `CachedChatCompletionClient` (a potential name) is a component that wraps `ChatCompletionClient` and provides the same behaviour .. just with caching.

**Comment by srjoglekar246** [comment]
`ChatCompletionCache` should indeed behave just like any other client. Do we need a separate class to just do `ChatCompletionCache(OpenAIChatCompletionClient(model="gpt-4o"))`? (I agree about the need for a default in-memory store, I can add that)

**Comment by ekzhu** [comment]
Let's make `ChatCompletionCache` having a default in-memory cache. It will be very simple.

**Comment by srjoglekar246** [comment]
https://github.com/microsoft/autogen/pull/5188
